lib/rubygems/package/tar_reader.rb

```diff
@@ -53,39 +53,11 @@ def close
   def each
     return enum_for __method__ unless block_given?
 
-    use_seek = @io.respond_to?(:seek)
-
     until @io.eof? do
       header = Gem::Package::TarHeader.from @io
       return if header.empty?
-
       entry = Gem::Package::TarReader::Entry.new header, @io
-      size = entry.header.size
-
       yield entry
-
-      skip = (512 - (size % 512)) % 512
-      pending = size - entry.bytes_read
-
-      if use_seek
-        begin
-          # avoid reading if the @io supports seeking
-          @io.seek pending, IO::SEEK_CUR
-          pending = 0
-        rescue Errno::EINVAL
-        end
-      end
-
-      # if seeking isn't supported or failed
-      while pending > 0 do
-        bytes_read = @io.read([pending, 4096].min).size
-        raise UnexpectedEOF if @io.eof?
-        pending -= bytes_read
-      end
-
-      @io.read skip # discard trailing zeros
-
-      # make sure nobody can use #read, #getc or #rewind anymore
       entry.close
```

Comment:
should this be in an `begin; yield entry; ensure; entry.close; end` block?

Reply:
It can't without making a breaking change because of the way each is exited during the iteration by seek (and because Enumerable is included). If I could make a breaking change I would ensure close.

Reply:
I think the solution is to remove Enumerable from TarReader since most of enumerable would produce broken already-closed entries. That's a notable breaking change on a minor part of rubygems though. Bumping a major or even minor version for it seems overzealous and leaving it as is probably won't hurt anyone.